pageserver/src/tenant/timeline.rs

```diff
@@ -2715,10 +2715,22 @@ impl Timeline {
     ) -> Result<HashMap<LayerFileName, LayerFileMetadata>, PageReconstructError> {
         let timer = self.metrics.create_images_time_histo.start_timer();
         let mut image_layers: Vec<ImageLayer> = Vec::new();
+
+        // We need to avoid holes between generated image layers.
+        // Otherwise LayerMap::image_layer_exists will return false if key range of some layer is covered by more than one
+        // image layer with hole between them. In this case such layer can not be utilized by GC.
+        //
+        // How such hole between partitions can appear?
+        // if we have relation with relid=1 and size 100 and relation with relid=2 with size 200 then result of
+        // KeySpace::partition may contain partitions <100000000..100000099> and <200000000..200000199>.
+        // If there is delta layer <100000000..300000000> then it never be garbage collected because
+        // image layers  <100000000..100000099> and <200000000..200000199> are not completely covering it.
+        let mut start_lsn = Key::MIN;
```

Comment:
Is the reason why we start from `Key::MIN` that it's the `DBDIR_KEY` instead of `partitioning.parts.first().map(|x| x.start).expect("should have at least one partitioning")`?

Reply:
Image layers should e as large as possible (in key range).
While for delta layers we are trying to exclude "holes" and split sparse delta  layers into several more dense ones,
image layers are "dense" by definition: them contain all available data for the specified key range.

This is the primary goal of this PR: avoid "holes" between delta layers which prevents utilisation of old layers by GC.
It is true for the first (starting) hole. But for it actually there is no difference, because Key::MIN = DBDIR_KEY is always present in the storage.

Reply:
Thanks for clarifying images vs. deltas. I think we are good to merge this!